### PR TITLE
fix: custom preset cached on target's preset.yml

### PR DESCRIPTION
**Pull request by fboucquez**
This is a small improvement. The original custom preset (without any private key) is cached in the preset.yml of the target folder. If the user upgrading doesn't provide the --customPreset, that cache is used. Currently, if the user is not providing a custom preset when --upgrading, a custom value (for example maxUnlockedAccounts: 20) could be reset to its bootstrap's default (10).

--preset and --assembly are already derived from the target folder when not provided. 

How to upgrade a node before:
```
symbol-bootstrap start --upgrade --preset mainnet --assembly dual --customPreset my-custom-preset.yml
```
How to upgrade a node after:
```
symbol-bootstrap start --upgrade
```
The custom preset is loaded from the cache inside the target folder, the same applies to the preset and assembly values. 
The user can still provide a custom preset if he wants to change a property for example. 

Note: User needs to run the full --upgrade command one more time so the custom preset cache is populated.

Another fix: The --preset bootstrap is not the default anymore. Most people are creating nodes for testnet and mainnet, not for local testing networks like it was originally intended. The preset must be explicitly defined when starting a new node.

### CODE REVIEW

**Comment by mm-s** [comment]
How does it behave when the custom preset contains private keys ?

**Comment by fboucquez** [comment]
> How does it behave when the custom preset contains private keys?

Private keys are not stored in the ./target/preset.yml file, this files contains the catched custom preset. 

You can create your custom preset in your local offline machine, generate the configuration, zip the target folder, and send it to your node without the node having the keys at all.  With this PR, you don't need to send the custom preset either.

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,8 @@ The changelog format is based on [Keep a Changelog](https://keepachangelog.com/e
 
 -   Added multi voting key file support.
 -   Added `updateVotingKeys` command.
+-   The `bootstrap` preset is not the default anymore. The name must be provided via --preset or as a custom preset field. 
+-   Added custom preset to the target folder. It's not required when upgrading the node without a configuration change.
```
Comment (segfaultxavi) [inline CHANGELOG.md:13]: It's **not** required? What do you mean?

File: src/service/ConfigLoader.ts [hunk 2]
```diff
@@ -383,13 +383,10 @@ export class ConfigLoader {
...
-        const preset =
-            params.preset ||
-            params.customPresetObject?.preset ||
-            customPresetFileObject?.preset ||
-            oldPresetData?.preset ||
-            Preset.bootstrap;
-
+        const preset = params.preset || params.customPresetObject?.preset || customPresetFileObject?.preset || oldPresetData?.preset;
+        if (!preset) {
+            throw new KnownError('Preset value could not be resolved. Have you provided the --preset parameter?');
```
Comment (segfaultxavi) [inline src/service/ConfigLoader.ts:388]:
```suggestion
            throw new KnownError('Preset value could not be resolved from target folder contents. Please provide the --preset parameter.');
```

**Review by segfaultxavi — CHANGES_REQUESTED** [review]
I understand the purpose of this PR and I like it very much. But I do not understand that sentence in the README.

File: CHANGELOG.md (hunk 1 above)
Comment (fboucquez) [reply CHANGELOG.md:13]: Soo, if you want to change a custom preset configuration, (like the friendly name), you need to provide the custom preset with --upgrade. If you are just upgrading the configuration, the custom preset is not required anymore. 

"The --customPreset param is only required when there is a configuration change, like a new 'friendlyName'"

File: src/service/ConfigLoader.ts [hunk 3]
```diff
@@ -383,13 +383,10 @@ export class ConfigLoader {
...
         const oldPresetData = params.oldPresetData;
         const customPresetFileObject = this.loadCustomPreset(customPreset, params.password);
-        const preset =
-            params.preset ||
-            params.customPresetObject?.preset ||
-            customPresetFileObject?.preset ||
-            oldPresetData?.preset ||
-            Preset.bootstrap;
-
+        const preset = params.preset || params.customPresetObject?.preset || customPresetFileObject?.preset || oldPresetData?.preset;
```
Comment (Wayonb) [inline src/service/ConfigLoader.ts:386]: Whats the order of precedence for the presets?  Should ``customPresetFileObject`` comes before ``customPresetObject``?

**Review by Wayonb — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (segfaultxavi) [reply CHANGELOG.md:13]: Ah, OK, understood. Then yeah, something like:

```suggestion
-   When upgrading, the --customPreset param is now only required if there is a configuration change, since
    the properties from the original custom preset are fully cached in the target folder.
```

But this is only the CHANGELOG so it's no big deal :D

**Review by segfaultxavi — DISMISSED** [review]

File: src/service/ConfigLoader.ts (hunk 3 above)
Comment (fboucquez) [reply src/service/ConfigLoader.ts:393]: if the user is providing both, a file and a custom preset object, the custom preset object is priority. This is the case when you use bootstrap as a library. 

I would probably need to revisit this, especially in private networks when the network.yml is provided instead of built-in bootstrap.

**Comment by fboucquez** [comment]
Superseded by https://github.com/nemtech/symbol-bootstrap/pull/277

**Comment by fboucquez** [comment]
Hi @Wayonb I'm keen to merge this one. Is it ok for you?

File: docs/config.md [hunk 4]
```diff
@@ -14,17 +14,19 @@ USAGE
   $ symbol-bootstrap config
 
 OPTIONS
-  -a, --assembly=assembly                   The assembly, example "dual" for testnet. If not provided, the value is
-                                            resolved from the target/preset.yml file.
+  -a, --assembly=assembly                   The assembly that define the node(s) layout. It can be provided via custom
```
Comment (Jaguar0625) [inline docs/config.md:17]: that define*s

Comment (Jaguar0625) [reply docs/config.md:17]: all three docs files. 

as there is a lot of similarity, i was wondering: are you generating these manually or automatically?

File: docs/start.md [hunk 5]
```diff
@@ -79,11 +80,10 @@ OPTIONS
 
   --upgrade
       It regenerates the configuration reusing the previous keys. Use this flag when upgrading the version of bootstrap to 
-      keep your node up to date without dropping the local data. The original preset (-t), assembly (-a), and custom 
-      preset (-a) must be used. Backup the target folder before upgrading.
+      keep your node up to date without dropping the local data. Backup the target folder before upgrading.
 
 EXAMPLES
-  $ symbol-bootstrap start
+  $ symbol-bootstrap start -p bootstrap
```
Comment (Jaguar0625) [inline docs/start.md:86]: delete? same as next line

File: README.md [hunk 6]
```diff
@@ -66,9 +66,9 @@ Properties in each file override the previous values (by object deep merge).
 
 ### Out-of-the-box presets:
 
--   `-p bootstrap`: Default [preset](https://github.com/nemtech/symbol-bootstrap/blob/main/presets/bootstrap/network.yml). It's a private network with 1 mongo database, 2 peers, 1 api and 1 rest gateway. Nemesis block is generated.
+-   `-p bootstrap`: A [preset](https://github.com/nemtech/symbol-bootstrap/blob/main/presets/bootstrap/network.yml). It's a private network with 1 mongo database, 2 peers, 1 api and 1 rest gateway. Nemesis block is generated.
```
Comment (Jaguar0625) [inline README.md:69]: shouldn't this be a named preset ?

File: src/commands/pack.ts [hunk 7]
```diff
@@ -112,7 +110,7 @@ export default class Pack extends Command {
         await ZipUtils.zip(targetZip, zipItems);
         await BootstrapUtils.deleteFile(noPrivateKeyTempFile);
         console.log();
-        console.log(`Zip file ${targetZip} has been created. You can unzip it in your node's machine and run:`);
-        console.log(`$ symbol-bootstrap start -p ${preset}${assembly ? ` -a ${assembly}` : ''} -c ${configOnlyCustomPresetFileName}`);
+        console.log(`You can unzip it in your node's machine and run:`);
```
Comment (Jaguar0625) [inline src/commands/pack.ts:113]: i think it would be nice to output location of zip still

File: src/commands/start.ts [hunk 8]
```diff
@@ -26,7 +25,7 @@ export default class Start extends Command {
     static description = 'Single command that aggregates config, compose and run in one line!';
 
     static examples = [
-        `$ symbol-bootstrap start`,
+        `$ symbol-bootstrap start -p bootstrap`,
         `$ symbol-bootstrap start -p bootstrap`,
```
Comment (Jaguar0625) [inline src/commands/start.ts:29]: remove duplicate line

File: src/commands/updateVotingKeys.ts [hunk 9]
```diff
@@ -57,16 +55,24 @@ When a new voting file is created, Bootstrap will advise running the \`link\` co
...
-        const preset = existingPreset.preset;
-        if (!preset) {
-            throw new Error(`Network preset could not be resolved!`);
-        }
-        // Adds new shared/network properties to the existing preset. This is for upgrades..
-        const sharedPreset = BootstrapUtils.loadYaml(join(BootstrapUtils.ROOT_FOLDER, 'presets', 'shared.yml'), false);
-        const networkPreset = BootstrapUtils.loadYaml(join(BootstrapUtils.ROOT_FOLDER, 'presets', preset, 'network.yml'), false);
-        const presetData = configLoader.mergePresets(sharedPreset, networkPreset, existingPreset) as ConfigPreset;
 
+        const loadPresetData = () => {
```
Comment (Jaguar0625) [inline src/commands/updateVotingKeys.ts:59]: why did you change this to a lambda instead of just being inline code?

File: src/service/ConfigLoader.ts [hunk 10]
```diff
@@ -124,8 +127,12 @@ export class ConfigLoader {
 
         if (presetData.nemesis) {
             if (oldAddresses) {
+                if (!oldPresetData) {
+                    throw new Error('oldPresetData must be defined when upgrading!');
+                }
                 // Nemesis configuration cannot be changed on upgrade.
                 addresses.mosaics = oldAddresses.mosaics;
+                presetData.nemesis = oldPresetData.nemesis;
```
Comment (Jaguar0625) [inline src/service/ConfigLoader.ts:135]: should you check that nemesis is the same ?

File: test/service/ConfigLoader.test.ts [hunk 11]
```diff
@@ -61,6 +63,114 @@ describe('ConfigLoader', () => {
         expect(presetData).to.not.be.undefined;
     });
 
+    it('ConfigLoader custom maxUnlockedAccounts', async () => {
```
Comment (Jaguar0625) [inline test/service/ConfigLoader.test.ts:66]: how is this test related to the rest of this change?

File: docs/config.md (hunk 4 above)
Comment (fboucquez) [reply docs/config.md:17]: Fixed. They are automatically generated

File: test/service/ConfigLoader.test.ts (hunk 11 above)
Comment (fboucquez) [reply test/service/ConfigLoader.test.ts:66]: It's testing that when recreating the `presetData` without providing the custom preset, the cached custom preset is used (and the user-provided maxUnlockedAccounts is kept for example).

File: src/commands/updateVotingKeys.ts (hunk 9 above)
Comment (fboucquez) [reply src/commands/updateVotingKeys.ts:59]: To avoid `*let* presetData: ConfigPreset;`, but it's an overkill. Lambda removed

File: README.md (hunk 6 above)
Comment (fboucquez) [reply README.md]: I've updated the docs for this pr. 

I'm planning to change the assemblies, please have a look at the draft pr https://github.com/symbol/symbol-bootstrap/pull/328

Preview doc https://github.com/symbol/symbol-bootstrap/tree/shared_assemblies#preset

File: src/commands/pack.ts (hunk 7 above)
Comment (fboucquez) [reply src/commands/pack.ts]: Added

File: src/commands/start.ts (hunk 8 above)
Comment (fboucquez) [reply src/commands/start.ts]: Removed

File: src/service/ConfigLoader.ts (hunk 10 above)
Comment (fboucquez) [reply src/service/ConfigLoader.ts]: I've removed that line. It's not currently required in this PR. 

I'm planning to provide a new PR improving how nemesis block is generated.
- Allow users to provide all the accounts, so the same custom preset executed twice (from scratch) generates the same accounts, balances, and nemesis block. 
- Allow users to include balances for both main and harvesting accounts (atm only the main mosaic can be 'opted-in')
- Validations that node's main accounts are not over the harvesting limit when dynamically generated (no user-provided)

File: src/service/ConfigLoader.ts (hunk 3 above)
Comment (Jaguar0625) [reply src/service/ConfigLoader.ts:398]: if you don't already, consider adding a build section to the readme (if you don't already) about how to generate these files and the expected test reporting files too. can be separate PR.

File: docs/config.md (hunk 4 above)
Comment (Jaguar0625) [reply docs/config.md]: if you don't already, consider adding a build section to the readme (if you don't already) about how to generate these files and the expected test reporting files too. can be separate PR.

**Review by Jaguar0625 — APPROVED** [review]
